File: practice/mathematics/fundamentals/find_point/FindPoint.js
function processData(input) {
    var lines = input.split("\n");
    //var points = parseInt(lines[0]);
    for(var i = 1; i < lines.length; i++) {
        var coordinates = lines[i].split(" ").map(i => parseInt(i));
        var x = coordinates[0];
        var y = coordinates[1];
        var mx = coordinates[2];
        var my = coordinates[3];
        var sx = 2 * mx - x;
        var sy = 2 * my - y;
        console.log(sx + " " + sy);
    }
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
    processData(_input);
});

File: practice/mathematics/fundamentals/find_point/FindPoint.scala
import scala.io.Source

case class Point(val x: Int, val y: Int) {
  def reflection(midPoint: Point): Point = {
    val rx = 2 * midPoint.x - x
    val ry = 2 * midPoint.y - y
    Point(rx, ry)
  }
}

object FindPoint extends App {
  val coordinates = Source.stdin.getLines().drop(1)
  coordinates.foreach {
    lines => {
      val ints = lines.split(" ").map(_.toInt)
      val point = Point(ints(0), ints(1))
      val midPoint = Point(ints(2), ints(3))
      val symmetricPoint = point.reflection(midPoint)
      println(s"${symmetricPoint.x} ${symmetricPoint.y}")
    }
  }
}
